clang/lib/Sema/SemaInit.cpp

```diff
@@ -10598,10 +10598,36 @@ QualType Sema::DeduceTemplateSpecializationFromInitializer(
   if (TemplateName.isDependent())
     return SubstAutoTypeDependent(TSInfo->getType());
 
-  // We can only perform deduction for class templates.
+  // We can only perform deduction for class templates or alias templates.
   auto *Template =
       dyn_cast_or_null<ClassTemplateDecl>(TemplateName.getAsTemplateDecl());
+  TemplateDecl* LookupTemplateDecl = Template;
+  if (!Template && getLangOpts().CPlusPlus20) { // type alias template
+    if (auto *AliasTemplate = dyn_cast_or_null<TypeAliasTemplateDecl>(
+            TemplateName.getAsTemplateDecl());
+        AliasTemplate) {
+      LookupTemplateDecl = AliasTemplate;
+      auto UnderlyingType = AliasTemplate->getTemplatedDecl()
+                                ->getUnderlyingType()
+                                .getDesugaredType(Context);
```

Comment:
Do we mean Desugar here, or Canonical?

Reply:
Here we'd like to get the underlying type with all alias stripped. I think both `Desugar` and `Canonical` can achieve the goal here, but `Desugar` seems more explicit.

Reply:
`Desugar` doesn't remove all of the typedefs/etc, and leaves sugar in pointees/references.  So this is very likely to end up being an underlying type of, for example, `SomeTypedef*`.  Canonical canonicalizes EVERYTHING.